### PR TITLE
PeerUpdates for refactored relays

### PR DESCRIPTION
**Pull request by mattkasun**
## Describe your changes
PeerUpdates for refactored relays

## Provide testing steps
set NET-137

## Checklist before requesting a review
- [ ] My changes affect only 10 files or less.
- [ ] I have performed a self-review of my code and tested it.
- [ ] If it is a new feature, I have added thorough tests, my code is <= 1450 lines.
- [ ] If it is a bugfix, my code is <= 200 lines.
- [ ] My functions are <= 80 lines.
- [ ] I have had my code reviewed by a peer.
- [ ] My unit tests pass locally.
- [ ] Netmaker is awesome.

### CODE REVIEW

File: mq/publishers.go [hunk 1]
```diff
@@ -108,6 +110,72 @@ func PublishSingleHostPeerUpdate(ctx context.Context, host *models.Host, deleted
...
+	for _, nodeI := range nodes {
+		peerHost, err := logic.GetHost(nodeI.HostID.String())
+		if err == nil {
+			publish(peerHost, fmt.Sprintf("peer/host/%s/%s", host.ID.String(), servercfg.GetServer()), data)
+		}
+	}
+	return nil
+}
+
+func BroadCastAddPeer(host *models.Host, node *models.Node, network string, update bool) error {
```
Comment (TobiaszCudnik) [inline mq/publishers.go:141]: What does `update` do?

File: mq/relay.go [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+	if err != nil {
+		logger.Log(0, "marshal peer update", err.Error())
+		return
+	}
+	fmt.Println("publishing peer update", client.Host.Name, p.Action, len(data))
+	publish(&client.Host, fmt.Sprintf("peer/host/%s/%s", client.Host.ID.String(), servercfg.GetServer()), data)
+}
+
+// pubRelayUpdate - publish peer update to a relay
+func pubRelayUpdate(client *models.Client, peers *[]models.Client) {
```
Comment (TobiaszCudnik) [inline mq/relay.go:250]: Pointer to a slice? Whats the reason?

File: mq/relay.go [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+		}
+		if peer.Node.IsRelay {
+			update.AllowedIPs = append(update.AllowedIPs, getAllowedIPs(peer)...)
+		}
+		if peer.Node.IsEgressGateway {
+			update.AllowedIPs = append(update.AllowedIPs, getEgressIPs(peer)...)
+		}
+		if peer.Node.IsIngressGateway {
+			update.AllowedIPs = append(update.AllowedIPs, getIngressIPs(peer)...)
+		}
```
Comment (TobiaszCudnik) [inline mq/relay.go:288]: Lines 272-287 could be extracted to a function, considering they appear 4 times in this file.

File: mq/publishers.go [hunk 4]
```diff
@@ -108,6 +110,72 @@ func PublishSingleHostPeerUpdate(ctx context.Context, host *models.Host, deleted
 	return publish(host, fmt.Sprintf("peers/host/%s/%s", host.ID.String(), servercfg.GetServer()), data)
 }
 
+func BroadCastDelPeer(host *models.Host, network string) error {
```
[the remark this answers is not in the document]
Comment (mattkasun) [reply mq/publishers.go:113]: due to rebase; not updating in this PR

File: mq/relay.go (hunk 2 above)
Comment (mattkasun) [reply mq/relay.go:250]: consistency so all parameter are passed by ref

File: mq/publishers.go (hunk 1 above)
Comment (mattkasun) [reply mq/publishers.go:141]: due to rebase; not updating in this PR

File: mq/publishers.go [hunk 5]
```diff
@@ -108,6 +110,72 @@ func PublishSingleHostPeerUpdate(ctx context.Context, host *models.Host, deleted
...
+		p.Action = models.UpdatePeer
+	}
+	for _, nodeI := range nodes {
+		// update allowed ips, according to the peer node
+		p.Peers[0].AllowedIPs = logic.GetAllowedIPs(&nodeI, node, nil)
+		data, err := json.Marshal(p)
+		if err != nil {
+			continue
+		}
+		peerHost, err := logic.GetHost(nodeI.HostID.String())
```
[the remark this answers is not in the document]
Comment (mattkasun) [reply mq/publishers.go:171]: due to rebase; not updating in this PR

File: mq/relay.go (hunk 2 above)
Comment (TobiaszCudnik) [reply mq/relay.go:250]: Slices and maps are always by ref in golang (besides metadata). Why a double pointer? It introduces unnecessary complexity, but maybe theres a benefit which is not clear from the code?

File: mq/relay.go (hunk 3 above)
Comment (mattkasun) [reply mq/relay.go]: function created

**Review by TobiaszCudnik — APPROVED** [review]

**Review by Aceix — APPROVED** [review]

**Pull request merged**